### PR TITLE
VM Dataloader and Optimiser rework to support multiple inputs

### PR DESCRIPTION
**Pull request by MissingNO57**

### CODE REVIEW

File: libs/vm_modules/src/ml/dataloaders/dataloader.cpp [hunk 1]
```diff
@@ -157,9 +158,30 @@ Ptr<VMTrainingPair> VMDataLoader::GetNext()
...
+  auto first = this->vm_->CreateNewObject<math::VMTensor>(next.first);
+
+  auto second = this->vm_->CreateNewObject<math::VMTensor>(next.second.at(0));
+
+  auto second_vector =
+      this->vm_->CreateNewObject<fetch::vm::Array<Ptr<fetch::vm_modules::math::VMTensor>>>(
+          second->GetTypeId(), static_cast<int32_t>(next.second.size()));
+
+  TemplateParameter1 element;
+  element.Assign(second, second->GetTypeId());
```
Comment (bipll) [inline libs/vm_modules/src/ml/dataloaders/dataloader.cpp]: Simply pass as construction arguments in the line above?

File: libs/vm_modules/src/ml/dataloaders/dataloader.cpp [hunk 2]
```diff
@@ -152,14 +153,33 @@ void VMDataLoader::AddTensorData(Ptr<VMTensorType> const &data, Ptr<VMTensorType
...
-  auto second     = this->vm_->CreateNewObject<math::VMTensor>(next.second.at(0));
-  auto dataHolder = this->vm_->CreateNewObject<VMTrainingPair>(first, second);
+  auto first  = this->vm_->CreateNewObject<math::VMTensor>(next.first);
+  auto second = this->vm_->CreateNewObject<math::VMTensor>(next.second.at(0));
+
+  auto second_vector =
+      this->vm_->CreateNewObject<fetch::vm::Array<Ptr<fetch::vm_modules::math::VMTensor>>>(
+          second->GetTypeId(), static_cast<int32_t>(next.second.size()));
+
+  TemplateParameter1 firt_element(second, second->GetTypeId());
```
Comment (evsmithx) [inline libs/vm_modules/src/ml/dataloaders/dataloader.cpp]: spelling

File: libs/vm_modules/src/ml/optimisation/optimiser.cpp [hunk 3]
```diff
@@ -108,14 +108,29 @@ void VMOptimiser::Bind(Module &module)
...
+    Ptr<String> const &label_node_name, Ptr<String> const &output_node_names)
 {
-  return new VMOptimiser(vm, type_id, mode->str, graph->GetGraph(), loader, {input_node_names->str},
+
+  std::vector<std::string> input_names;
+
+  AnyInteger         first_index(0, TypeIds::UInt16);
+  TemplateParameter1 element = input_node_names->GetIndexedValue(first_index);
+  input_names.push_back(element.Get<fetch::vm::Ptr<fetch::vm::String>>()->str);
+  fetch::math::SizeType cnt = input_names.size();
```
Comment (evsmithx) [inline libs/vm_modules/src/ml/optimisation/optimiser.cpp]: Shouldn't `cnt` be the length of `input_node_names` and not `input_names`? As I understand the code here, cnt is set to 1 and then the loop below never executes. Also, can we rename this variable please?! I keep reading it as c*nt :-P.

**Review by evsmithx — CHANGES_REQUESTED** [review]

**Review by evsmithx — DISMISSED** [review]
LGTM

**Review by evsmithx — APPROVED** [review]
The static analysis is really petty!

**Pull request merged**